src/lexer.py

```diff
@@ -0,0 +1,1060 @@
+TT_RPAREN   	= 'RPAREN'
+TT_EE			= 'EE'
+TT_NE			= 'NE'
+TT_LT			= 'LT'
+TT_GT			= 'GT'
+TT_LTE			= 'LTE'
+TT_GTE			= 'GTE'
+TT_EOF			= 'EOF'
+
+KEYWORDS = [
+	'var','and','or','not',
+	'if','elif','else','for',
+	'to','step','while','then'
+]
+
+class Token:
+	def __init__(self, type_, value=None, pos_start=None, pos_end=None):
+		self.type = type_
+		self.value = value
+
+		if pos_start:
+			self.pos_start = pos_start.copy()
+			self.pos_end = pos_start.copy()
+			self.pos_end.advance()
+
+		if pos_end:
+			self.pos_end = pos_end.copy()
+
+	def matches(self, type_, value):
+		return self.type == type_ and self.value == value
+
+	def __repr__(self):
+		if self.value: return f'{self.type}:{self.value}'
+		return f'{self.type}'
+
+#######################################
+# LEXER
+#######################################
+
+class Lexer:
+	def __init__(self, fn, text):
+		self.fn = fn
+		self.text = text
+		self.pos = Position(-1, 0, -1, fn, text)
+		self.current_char = None
+		self.advance()
+
+	def advance(self):
+		self.pos.advance(self.current_char)
+		self.current_char = self.text[self.pos.idx] if self.pos.idx < len(self.text) else None
+
+	def make_tokens(self):
+		tokens = []
+
+		while self.current_char != None:
+			if self.current_char in ' \t':
+				self.advance()
+			elif self.current_char in DIGITS:
+				tokens.append(self.make_number())
+			elif self.current_char in LETTERS:
+				tokens.append(self.make_identifier())
+			elif self.current_char == '+':
+				tokens.append(Token(TT_PLUS, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '-':
+				tokens.append(Token(TT_MINUS, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '*':
+				tokens.append(Token(TT_MUL, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '/':
+				tokens.append(Token(TT_DIV, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '^':
+				tokens.append(Token(TT_POW, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '(':
+				tokens.append(Token(TT_LPAREN, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == ')':
+				tokens.append(Token(TT_RPAREN, pos_start=self.pos))
+				self.advance()
+			elif self.current_char == '!':
+				token, error = self.make_not_equals()
+				if error: return [], error
+				tokens.append(token)
+			elif self.current_char == '=':
+				tokens.append(self.make_equals())
+			elif self.current_char == '<':
+				tokens.append(self.make_less_than())
+			elif self.current_char == '>':
+				tokens.append(self.make_greater_than())
+			else:
+				pos_start = self.pos.copy()
+				char = self.current_char
+				self.advance()
+				return [], IllegalCharError(pos_start, self.pos, "'" + char + "'")
+
+		tokens.append(Token(TT_EOF, pos_start=self.pos))
+		return tokens, None
+
+	def make_number(self):
+		num_str = ''
+		dot_count = 0
+		pos_start = self.pos.copy()
+
+		while self.current_char != None and self.current_char in DIGITS + '.':

```

Comment:
Additionally, I'm not sure what's happening in the second part of this "and", but some parenthesis would help readability. Also, a comment here would be suggested.